docs/proposals/20250513-chained-and-efficient-upgrades-for-clusters-with-managed-topologies.md

```diff
@@ -0,0 +1,408 @@
+the same machinery that will be used during upgrades. 
+
+That means that by doing any rollout, e.g. due to an automatic machine remediation, you get a proxy signal about the
+fact that the system can successfully perform an upgrade, or you get the chance to detect and fix issues in the system
+before a full upgrade is performed.
+
+Conversely, risk increase for users not performing any form of rollouts for long periods.
+
+- Upgrading a Cluster by multiple Kubernetes minor versions might compromise workloads.
+
+Each Kubernetes minor upgrade might include changes impacting workloads running in the Cluster, and going through
+multiple minors increases the chance that workloads running in a cluster are impacted.
+
+As a mitigation, this proposal is improving lifecycle hooks and giving full visibility in the upgrade plan, thus
+allowing users to implement additional pre-upgrade checks.
+
+## Alternatives
+
+An alternative to option leveraging a new CRs to define the list of Kubernetes version to be used for upgrade plans
+was considered.
+
+However, the option was discarded because it seems more consistent having the list of 
+Kubernetes version to be used for upgrade plans in ClusterClasses, alongside all the other info defining 
+how a managed topology should behave.  
+
+## Upgrade Strategy
+
+No particular upgrade considerations are required, this feature will available to users upgrading to
+Cluster API v1.11.
+
+However, it is required to enhance ClusterClasses with the information required to compute upgrade plans,
+otherwise the system will keep supporting only upgrade to the next minor for the corresponding clusters (opt-in). 
+
+## Additional Details
+
+### Test Plan
+
+Considering the entire range of [Kubernetes versions currently supported by Cluster API](https://cluster-api.sigs.k8s.io/reference/versions#kubernetes-versions-support) 
+plus the Kubernetes version currently under development (from N-6 to N+1), more than twenty additional E2E tests are 
+required to get full coverage of the possible chained upgrade sequences. 
+
+While implementing all those new tests is not impossible, it is considered not practical because the resulting E2E
+job would take a long time while current E2E jobs allow a fast iterative development process.
+
+Accordingly, int the first iteration only one chained upgrade test scenario going from N-3 to N+1 will be validated, 
+but this is considered enough to ensure that:
+- The mechanics for chained upgrade works
+- [Kubernetes version skew policy](https://kubernetes.io/releases/version-skew-policy/) is respected, and workers upgrade are performed only when necessary
+- Lifecycle hooks are called
+- Resulting K8s cluster pass the conformance test
+
+This new test will run periodically, and also be available to be run on demand on PRs.
+
+### Graduation Criteria
+
+This feature should be considered part of ClusterClass and managed topologies, and thus it will follow the same graduation process.
```

Comment:
I think we should consider adding an additional feature gate and use it to guard the controller and webhook logic.

We were very close to enable ClusterClass per default. If we include this in the ClusterClass webhook we have to promote everything at the same time (including Runtime SDK and maybe even lifecycle hooks?)

Reply:
TBH I think we should keep things as simple, I'm not sure how many users understand nuances here.
Let's chat about it